### PR TITLE
Performance improvements

### LINKED ISSUE #7371: Slow loading speeds
**Describe the bug**
For our project (that contains quite a lot of metrics) the loading speeds of various pages are quite long. The main overview page almost always takes more than 10 seconds to load, which seems quite slow. It makes navigating through the pages quite frustrating. 
It seems that for the main overview (which does not show a whole lot of information) a lot of data is queried, leading to a lot of CPU usage on the database.

**Expected behaviour**
Loading times are more acceptable (at most a few seconds).

**Your environment (please complete the following information):**
 - Running on ICTU ECO 2 (NLA project)
 - Quality-time version 5.1

### PR DESCRIPTION
**Pull request by fniessink**
- In the /api/v3/report endpoint, get the recent measurements for all metrics in one database call, instead of using one database call per report.
- Create a measurements index on both the start and the end field instead of just on the start field.
- Create an index on the report_uuid field of the reports to speed up the `distinct` query.

### CODE REVIEW

File: components/api_server/src/initialization/database.py [hunk 1]
```diff
@@ -42,9 +42,9 @@ def set_feature_compatibility_version(admin_database: Database) -> None:
 def create_indexes(database: Database) -> None:
     """Create any indexes."""
     database.datamodels.create_index("timestamp")
-    database.reports.create_index("timestamp")
+    database.reports.create_index([("timestamp", pymongo.ASCENDING), ("report_uuid", pymongo.ASCENDING)])
     database.users.create_index("username", unique=True)
-    start_index = pymongo.IndexModel([("start", pymongo.ASCENDING)])
+    start_index = pymongo.IndexModel([("start", pymongo.ASCENDING), ("end", pymongo.DESCENDING)])
```
Comment (wkoot) [inline components/api_server/src/initialization/database.py]: Is the name `start_index` still accurate?

File: components/api_server/src/database/reports.py [hunk 2]
```diff
@@ -81,7 +81,7 @@ def latest_reports(database: Database) -> list[Report]:
 
 def latest_reports_before_timestamp(database: Database, data_model: dict, max_iso_timestamp: str) -> list[Report]:
     """Return the latest, undeleted, reports in the reports collection before the max timestamp."""
-    report_filter = {"timestamp": {"$lt": max_iso_timestamp}} if max_iso_timestamp else {}
+    report_filter = {"timestamp": {"$lte": max_iso_timestamp}} if max_iso_timestamp else {}
```
Comment (wkoot) [inline components/api_server/src/database/reports.py:84]: Why the addition of equal? Was the max timestamp not something very far in the future?

File: tests/feature_tests/src/steps/report.py [hunk 3]
```diff
@@ -131,13 +131,12 @@ def import_failed(context: Context) -> None:
...
-def get_measurements(context: Context, has_or_had: str, expected_number: str) -> None:
+@then('the report has "{expected_number}" measurements')
+def get_measurements(context: Context, expected_number: str) -> None:
     """Get the recent measurements of a report."""
-    if has_or_had == "had":
-        context.report_date = "2020-11-17T10:00:00Z"
-        context.min_report_date = "2020-11-16T00:00:00Z"
-
+    now = datetime.now(tz=UTC).replace(microsecond=0)
+    context.report_date = (now + timedelta(days=10)).isoformat()
```
Comment (wkoot) [inline tests/feature_tests/src/steps/report.py:138]: Why is this needed?

File: components/api_server/src/database/measurements.py [hunk 4]
```diff
@@ -11,12 +11,21 @@
 from shared.utils.functions import iso_timestamp
 from shared.utils.type import MetricId
 
+# Projection options
+NO_ID = {"_id": False}
+NO_SOURCE_DETAILS = NO_ID | {"sources.entities": False, "sources.entity_user_data": False}
+NO_MEASUREMENT_DETAILS = NO_SOURCE_DETAILS | {"issue_status": False}
+
+# Sort optins
+START_ASCENDING = [("start", pymongo.ASCENDING), ("end", pymongo.DESCENDING)]
```
Comment (wkoot) [inline components/api_server/src/database/measurements.py:20]: Why does this also contain "end"?

File: components/api_server/src/database/measurements.py [hunk 5]
```diff
@@ -27,52 +36,24 @@ def count_measurements(database: Database) -> int:
...
-                "_id": False,
-                "sources.entities": False,
-                "sources.entity_user_data": False,
-                "issue_status": False,
-            },
-        ),
-    )
+def measurements_in_period(database: Database, min_iso_timestamp: str, max_iso_timestamp: str):
+    """Return recent measurements for the specified metrics, without source details and issue status."""
+    measurement_filter = {"start": {"$lte": max_iso_timestamp or iso_timestamp()}, "end": {"$gte": min_iso_timestamp}}
```
Comment (wkoot) [inline components/api_server/src/database/measurements.py:43]: I don't get this; start on less-than-or-equals-to MAX, end on greater-than-or-equals to MIN, but sort ascending?
I would expect either the start/end to be reversed, or the sorting to be descending. Perhaps describe/comment?

File: components/api_server/src/database/measurements.py [hunk 6]
```diff
@@ -27,52 +36,24 @@ def count_measurements(database: Database) -> int:
...
     if max_iso_timestamp:
-        measurement_filter["start"] = {"$lt": max_iso_timestamp}
-    latest_measurement_complete = database.measurements.find_one(
-        measurement_filter,
-        sort=[("start", pymongo.DESCENDING)],
-        projection={"_id": False},
-    )
-    if not latest_measurement_complete:
+        measurement_filter["start"] = {"$lte": max_iso_timestamp}
+    latest_measurement = database.measurements.find_one(measurement_filter, sort=START_ASCENDING, projection=NO_ID)
```
Comment (wkoot) [inline components/api_server/src/database/measurements.py:52]: Why do we use the projection?

File: components/api_server/src/database/measurements.py [hunk 7]
```diff
@@ -27,52 +36,24 @@ def count_measurements(database: Database) -> int:
...
+    if not latest_measurement:
         return []
-    all_measurements_stripped = measurements_by_metric(database, metric_uuid, max_iso_timestamp=max_iso_timestamp)
-    return list(all_measurements_stripped)[:-1] + [latest_measurement_complete]
+    all_measurements_stripped = list(
+        database.measurements.find(measurement_filter, sort=START_ASCENDING, projection=NO_MEASUREMENT_DETAILS),
+    )
+    return list(all_measurements_stripped)[:-1] + [latest_measurement]
```
Comment (wkoot) [inline components/api_server/src/database/measurements.py]: The `list` cast is duplicate

File: components/api_server/src/database/measurements.py [hunk 8]
```diff
@@ -27,52 +36,24 @@ def count_measurements(database: Database) -> int:
...
-        projection={"_id": False},
-    )
-    if not latest_measurement_complete:
+        measurement_filter["start"] = {"$lte": max_iso_timestamp}
+    latest_measurement = database.measurements.find_one(measurement_filter, sort=START_ASCENDING, projection=NO_ID)
+    if not latest_measurement:
         return []
-    all_measurements_stripped = measurements_by_metric(database, metric_uuid, max_iso_timestamp=max_iso_timestamp)
-    return list(all_measurements_stripped)[:-1] + [latest_measurement_complete]
+    all_measurements_stripped = list(
```
Comment (wkoot) [inline components/api_server/src/database/measurements.py:55]: Would it be a performance improvement to yield a generator instead, i.e. if we don't iterate through the *entire* history? Although that would make more sense in desc sort

File: components/api_server/src/database/measurements.py [hunk 9]
```diff
@@ -81,21 +62,15 @@ def recent_measurements(
...
-    )
+    measurement_filter: dict[str, Any] = {"start": {"$lte": max_iso_timestamp}, "end": {"$gte": min_iso_timestamp}}
+    measurements = database.measurements.find(measurement_filter, sort=START_ASCENDING, projection=NO_SOURCE_DETAILS)
     measurements_by_metric_uuid: dict[MetricId, list] = {}
     for measurement_dict in measurements:
         metric_uuid = measurement_dict["metric_uuid"]
-        metric = metrics_dict[metric_uuid]
-        measurement = Measurement(metric, measurement_dict)
-        measurements_by_metric_uuid.setdefault(metric_uuid, []).append(measurement)
+        if metric := metrics_dict.get(metric_uuid):  # pragma: no feature-test-cover
```
Comment (wkoot) [inline components/api_server/src/database/measurements.py:76]: All calling code can handle getting an empty result `if not metric`?

File: components/api_server/src/initialization/database.py (hunk 1 above)
Comment (fniessink) [reply components/api_server/src/initialization/database.py]: Hmm, it's still an index on start first, but maybe `period_index` is better.

File: components/api_server/src/database/reports.py (hunk 2 above)
Comment (fniessink) [reply components/api_server/src/database/reports.py:84]: I got a failing testcase where the max_iso_timestamp was equal to the current time. This fixed it. I also think the less-than-or-equal is more intuitive in the case of timestamps.

File: components/api_server/src/database/measurements.py (hunk 4 above)
Comment (fniessink) [reply components/api_server/src/database/measurements.py:20]: So that it matches the index, see first alinea here: https://www.mongodb.com/docs/manual/tutorial/sort-results-with-indexes/#std-label-sorting-with-indexes

File: tests/feature_tests/src/steps/report.py (hunk 3 above)
Comment (fniessink) [reply tests/feature_tests/src/steps/report.py:138]: Because the endpoint now always needs a period (`min_report_date` and `report_date`). I tried to get rid of optional parameters that were basically only optional for sake of the tests.

File: components/api_server/src/database/measurements.py (hunk 5 above)
Comment (fniessink) [reply components/api_server/src/database/measurements.py:43]: Comment added:

```python
    # Return measurements that partially or completely overlap with the period, meaning they have their start before
    # the end of the period (max_iso_timestamp) and their end after the start of the period (min_iso_timestamp):
```

File: components/api_server/src/database/measurements.py (hunk 6 above)
Comment (fniessink) [reply components/api_server/src/database/measurements.py:52]: In general, to reduce data sent over the wire. Why the id is removed in particular I don't recall.

File: components/api_server/src/database/measurements.py (hunk 8 above)
Comment (fniessink) [reply components/api_server/src/database/measurements.py:55]: I don't think it matters, all data is sent to the frontend anyway.

File: components/api_server/src/database/measurements.py (hunk 9 above)
Comment (fniessink) [reply components/api_server/src/database/measurements.py:76]: The `if` is needed because there may be measurements for metrics that have been deleted. I'll add a comment.

File: tests/feature_tests/src/steps/report.py (hunk 3 above)
Comment (wkoot) [reply tests/feature_tests/src/steps/report.py:138]: I was wondering about the +10 days in particular :-) But shouldn't be a problem
